godot-core/src/builtin/vector_macros.rs

```diff
@@ -0,0 +1,275 @@
+    (
+        // Name of the vector type.
+        $Vector:ty,
+        // Type of each individual component, for example `i32`.
+        $Scalar:ty,
+        // Names of the components, with parentheses, for example `(x, y)`.
+        ($($components:ident),*),
+        // Name of the operator trait, for example `Add`.
+        $Operator:ident,
+        // Name of the function on the operator trait, for example `add`.
+        $func:ident$(,)?
+    ) => {
+        impl std::ops::$Operator<$Vector> for $Scalar {
+            type Output = $Vector;
+            fn $func(self, mut rhs: $Vector) -> Self::Output {
+                $(
+                    rhs.$components = rhs.$components.$func(self);
+                )*
+                rhs
+            }
+        }
+    }
+}
+
+/// Implements a single arithmetic assignment operator for a vector type, with a vector on the
+/// right-hand side.
+macro_rules! impl_vector_vector_assign_operator {
+    (
+        // Name of the vector type.
+        $Vector:ty,
+        // Type of each individual component, for example `i32`.
+        $Scalar:ty,
+        // Names of the components, with parentheses, for example `(x, y)`.
+        ($($components:ident),*),
+        // Name of the operator trait, for example `AddAssign`.
+        $Operator:ident,
+        // Name of the function on the operator trait, for example `add_assign`.
+        $func:ident$(,)?
+    ) => {
+        impl std::ops::$Operator for $Vector {
+            fn $func(&mut self, rhs: $Vector) {
+                $(
+                    self.$components.$func(rhs.$components);
+                )*
+            }
+        }
+    }
+}
+
+/// Implements a single arithmetic assignment operator for a vector type, with a scalar on the
+/// right-hand side.
+macro_rules! impl_vector_scalar_assign_operator {
+    (
+        // Name of the vector type.
+        $Vector:ty,
+        // Type of each individual component, for example `i32`.
+        $Scalar:ty,
+        // Names of the components, with parentheses, for example `(x, y)`.
+        ($($components:ident),*),
+        // Name of the operator trait, for example `AddAssign`.
+        $Operator:ident,
+        // Name of the function on the operator trait, for example `add_assign`.
+        $func:ident$(,)?
+    ) => {
+        impl std::ops::$Operator<$Scalar> for $Vector {
+            fn $func(&mut self, rhs: $Scalar) {
+                $(
+                    self.$components.$func(rhs);
+                )*
+            }
+        }
+    }
+}
+
+/// Implements all common arithmetic operators on a built-in vector type.
+macro_rules! impl_vector_operators {
+    (
+        // Name of the vector type to be implemented, for example `Vector2`.
+        $Vector:ty,
+        // Type of each individual component, for example `f32`.
+        $Scalar:ty,
+        // Names of the components, with parentheses, for example `(x, y)`.
+        ($($components:ident),*)$(,)?
```

Comment:
The trailing `$(,)?` is not really needed as this macro is not user-facing, as such we can control the call site.

Reply:
Done.